File: app/src/main/java/com/example/healthtracker/feature/listCalc/presentation/ListCalcPresenter.java
package com.example.healthtracker.feature.listCalc.presentation;

import com.example.healthtracker.feature.listCalc.ListCalc;
import com.example.healthtracker.model.Calc;
import com.example.healthtracker.model.CalcDao;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ListCalcPresenter implements ListCalc.Presenter {

    private ListCalc.View view;
    private final ExecutorService executorService;

    public ListCalcPresenter(ListCalc.View view) {
        this.view = view;
        this.executorService = Executors.newSingleThreadExecutor();
    }

    @Override
    public void getAllRegisters(CalcDao dao, String type) {
        executorService.execute(() -> {
            List<Calc> list = dao.getRegisterByType(type);
            if (view != null) {
                new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> {
                    if (view != null) {
                        view.displayAllRegisters(list);
                    }
                });
            }
        });
    }

    @Override
    public void clearRegisters(List<Calc> list, CalcDao dao, String type) {
        executorService.execute(() -> {
            if (list.isEmpty()) {
                if (view != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> {
                        if (view != null) {
                            view.displayFailure("");
                        }
                    });
                }
            } else {
                dao.deleteAllByType(type);
                if (view != null) {
                    new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> {
                        if (view != null) {
                            view.onDeleteRegisters();
                        }
                    });
                }
            }
        });
    }

    @Override
    public void onDestroy() {
        if (!executorService.isShutdown()) {
            executorService.shutdown();
        }
        view = null;
    }
}

File: app/src/main/java/com/example/healthtracker/feature/listCalc/presentation/ListCalcPresenter.kt
package com.example.healthtracker.feature.listCalc.presentation

import android.content.Context
import android.content.res.Resources
import com.example.healthtracker.R
import com.example.healthtracker.feature.listCalc.ListCalc
import com.example.healthtracker.model.Calc
import com.example.healthtracker.model.CalcDao
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ListCalcPresenter(private var view: ListCalc.View?): ListCalc.Presenter {

    private val presenterScope = CoroutineScope(Dispatchers.IO)

    override fun getAllRegisters(dao: CalcDao, type: String){
        presenterScope.launch {
            val list = dao.getRegisterByType(type)
            withContext(Dispatchers.Main){
                view?.displayAllRegisters(list)
            }
        }
    }

    override fun clearRegisters(list: List<Calc>, dao: CalcDao, type: String) {
        presenterScope.launch {
            if (list.isEmpty()){
                withContext(Dispatchers.Main){
                    view?.displayFailure("")
                }
            } else {
                dao.deleteAllByType(type)
                withContext(Dispatchers.Main){
                    view?.onDeleteRegisters()
                }
            }
        }
    }

    override fun onDestroy() {
        presenterScope.cancel()
        view = null
    }

}